tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/FunctionEmitter.scala

```diff
@@ -199,41 +199,100 @@ private[emitter] class FunctionEmitter(jsGen: JSGen) {
   }
 
   /** Desugars parameters and body to a JS function.
    */
-  def desugarToFunction(
-      params: List[ParamDef],
-      body: Tree, isStat: Boolean)(
-      implicit globalKnowledge: GlobalKnowledge, pos: Position): js.Function = {
+  def desugarToFunction(params: List[ParamDef], body: Tree, isStat: Boolean)(
+      implicit globalKnowledge: GlobalKnowledge,
+      pos: Position): WithGlobals[js.Function] = {
     new JSDesugar().desugarToFunction(params, body, isStat, Env.empty)
   }
 
   private class JSDesugar()(implicit globalKnowledge: GlobalKnowledge) {
 
-    // Synthetic variables
+    // Name management
+
+    /** Whether we are running in the "optimistic naming" run.
+     *
+     *  In theory, `JSDesugar` works in two passes: the optimistic run,
+     *  followed by the pessimistic run should the first one fail.
+     *
+     *  The optimistic run assumes that there is no clash between a local
+     *  variable name and a global variable name (i.e., a `JSGlobalRef`). This
+     *  allows it to straightforwardly translate IR identifiers to JS
+     *  identifiers. While it does that, it records all the local variable and
+     *  global variable names that are used in the method.
+     *
+     *  At the end of the translation, we check whether there was a clash by
+     *  testing if the two sets intersect. If they do not, we are lucky, and
+     *  can completely by-pass the pessimistic run. If there is a clash, then
+     *  we need to restart everything in pessimistic mode.
+     *
+     *  In the pessimistic run, we use the set of global variable names that
+     *  was collected during the optimistic run to *prevent* clashes from
+     *  happening. This requires that we maintain a map of IR identifiers to
+     *  allocated JS identifiers, as some IR identifiers need to be renamed.
+     */
+    private var isOptimisticNamingRun: Boolean = true
+
+    private val globalVarNames = mutable.Set.empty[String]
+    private val localVarNames = mutable.Set.empty[String]
+
+    private lazy val localVarAllocs = mutable.Map.empty[String, String]
+
+    private def referenceGlobalName(name: String): Unit =
+      globalVarNames += name
+
+    private def extractWithGlobals[A](withGlobals: WithGlobals[A]): A = {
+      for (globalRef <- withGlobals.globalVarNames)
+        referenceGlobalName(globalRef)
+      withGlobals.value
+    }
+
+    private def transformLocalName(name: String): String = {
+      if (isOptimisticNamingRun) {
+        localVarNames += name
+        name
+      } else {
+        def slowPath(): String = {
+          localVarAllocs.getOrElseUpdate(name, {
+            var suffix = 0
+            var result = name
+            while (globalVarNames.contains(result) ||
+                localVarNames.contains(result)) {
+              suffix += 1
+              result = name + "$" + suffix
+            }
+            localVarNames += result
+            result
```

Comment:
Although, I think here it's not a big deal. It's quadratic in theory, but in practice you need to try very hard to expose a quadratic behavior.

First, `showPath()` is only called in the pessimistic run. So like 1 in 40,000 methods. OK, that's not a valid explanation, since the `freshName()` problem in the optimizer happened for only 1 method too.

The real explanation is that we're not creating an arbitrary large number of fresh names for the same `base` (here called `name`). We only create one "fresh" name for any given base `name`. The optimizer problem was quadratic because we created thousands of fresh names from the same `base`. Here that cannot happen. Here, if we loop more than once, it's because there is *another* name *in the input* that happens to have `name$1` has its original name. And if that's the case, then creating a fresh name for `name$1` will only loop if there is a `name$1$1`. Which means that, for each iteration of the loop for any given input name `nameFoo`, there must also be an input `nameFoo$1`. I.e., the iterations for `nameFoo` are actually already amortized by the existence of `nameFoo$1` in the input.

So, after all, I believe the complexity of `slowPath()` is amortized O(1).